components/scream/src/share/util/scream_kokkos_utils.hpp

```diff
@@ -117,61 +119,136 @@ struct ExeSpaceUtils<Kokkos::Cuda> {
 
 /*
  * TeamUtils contains utilities for getting concurrency info for
- * thread teams.
+ * thread teams. Don't use _TeamUtilsCommonBase directly, use
+ * TeamUtils.
  */
+
 template <typename ExeSpace = Kokkos::DefaultExecutionSpace>
-class TeamUtils
+class _TeamUtilsCommonBase
 {
-  int _team_size, _num_teams;
+ protected:
+  int _team_size, _num_teams, _max_threads;
 
-public:
+ public:
   template <typename TeamPolicy>
-  TeamUtils(const TeamPolicy& policy) : _team_size(0)
+  _TeamUtilsCommonBase(const TeamPolicy& policy) : _team_size(0)
   {
-    const int max_threads = ExeSpace::concurrency();
+    _max_threads = ExeSpace::concurrency() / ( (!is_single_precision<Real>::value && OnGpu<ExeSpace>::value) ? 2 : 1);
```

Comment:
Why divide by 2 in dp on GPU?

Reply:
Number of GPU threads in Single-prec vs. Double-prec

Reply:
Is the `team_size` the same in SP and DP though? I think that the number of teams should not change in SP and DP. How you use the threads in a warp, does not change (I think) how many teams you have. So I think one should always divide by 2, even in sp.

Reply:
I'm not convinced unless concurrency() is somehow returning `80*32*32` for DP and `80*64*32` for SP. Another hypothesis is that we need to slightly overprovide workspaces to reduce lock contention.

Reply:
Concurrency must return the same value, or else I claim black magic here.

Re: nteams in sp and dp, I am not sure how CUDA implements the 2sp threads within a single thread. I was assuming that within the same kernel, you could use the same thread as 1dp thread or 2sp thread, and that had no effect with how many warps/blocks can concurrently run...